flax/linen/fp8_ops.py

```diff
@@ -0,0 +1,193 @@
+  sf = jnp.where(amax > 0.0, sf, scale)
+  sf = jnp.where(lax.is_finite(amax), sf, scale)
+  sf = jnp.where(exp < 0, 1.0 / sf, sf)
+  return 1.0 / sf
+
+def compute_scale_and_amax_history(x, q_dtype, scale, amax_history):
+  dtype_max = get_fp8_max(q_dtype, jnp.float32)
+
+  amax_update = jnp.max(jnp.abs(x)).astype(scale.dtype)
+  new_amax_history = \
+      jnp.roll(amax_history, shift=-1, axis=0).at[0].set(amax_update)
+
+  amax_from_history = jnp.max(new_amax_history, axis=0)
+  new_scale = compute_scale(amax_from_history, scale, dtype_max)
+  return new_scale, new_amax_history
+
+def qdq_and_return(x, q_dtype, scale, amax_history, compute_dtype):
+  qx = quantize_dequantize(x, q_dtype, scale, compute_dtype)
+  new_scale, new_amax_history = compute_scale_and_amax_history(
+      x, q_dtype, scale, amax_history)
+  return qx, new_scale, new_amax_history
+
+@partial(custom_vjp, nondiff_argnums=(0,))
+def in_qdq(compute_dtype, inp, scale, amax_history):
+  qin, _, _ = qdq_and_return(
+      inp, jnp.float8_e4m3fn, scale, amax_history, compute_dtype)
+  return qin
+
+def in_qdq_fwd(compute_dtype, inp, scale, amax_history):
+  qin, new_scale, new_amax_history = qdq_and_return(
+      inp, jnp.float8_e4m3fn, scale, amax_history, compute_dtype)
+  return qin, (new_scale, new_amax_history)
+
+def in_qdq_bwd(compute_dtype, res, g):
+  new_scale, new_amax_history = res
+  q_g = g
+  return q_g, new_scale, new_amax_history
+
+in_qdq.defvjp(in_qdq_fwd, in_qdq_bwd)
+
+
+@partial(custom_vjp, nondiff_argnums=(0,))
+def out_qdq(compute_dtype, out, scale, amax_history):
+  return out
+
+def out_qdq_fwd(compute_dtype, out, scale, amax_history):
+  return out, (scale, amax_history)
+
+def out_qdq_bwd(compute_dtype, res, g):
+  scale, amax_history = res
+  q_g, new_scale, new_amax_history = qdq_and_return(
+      g, jnp.float8_e5m2, scale, amax_history, compute_dtype)
+  return q_g, new_scale, new_amax_history
+  
+out_qdq.defvjp(out_qdq_fwd, out_qdq_bwd)
+
+def fp8_dot_general(lhs, rhs, dimension_numbers, precision, compute_dtype,
+                    lhs_scale, lhs_amax_history, rhs_scale, rhs_amax_history,
+                    dout_scale, dout_amax_history):
+  """Perform dot_general.  """
+  
+  lhs_qdq = in_qdq(compute_dtype, lhs, lhs_scale, lhs_amax_history)
+
+  rhs_qdq = in_qdq(compute_dtype, rhs, rhs_scale, rhs_amax_history)
+
+  output_qdq = lax.dot_general(lhs_qdq, rhs_qdq, dimension_numbers, precision)
+
+  out = out_qdq(compute_dtype, output_qdq, dout_scale, dout_amax_history)
+
+  return out
+
+
+class Fp8DenseGeneralOp(Module):
+  amax_history_length: int = 1024
+
+  def setup(self) -> None:
+    scale_args = (
+        initializers.ones_init(),
+        random.PRNGKey(0),
+        (1,),
+        jnp.float32,
+    )
+    amax_history_args = (
+        initializers.zeros_init(),
+        random.PRNGKey(0),
+        (self.amax_history_length,),
+        jnp.float32,
+    )
+
+    self.input_amax_history = variable_with_axes(
```

Comment:
...why are you using the deprecated paritioning helper `variable_with_axes` to seed a trivial piece of metadata here?  Why not just use `self.variable`?  I don't see anything in this code that's actually doing any non-trivial operation using this metadata.

Reply:
Sure. I think we can simply use the `self.variable`. The reason why we used the `variable_with_axes`  was that we originally developed this in the context of t5x, where it requires all the variables have the axes information. Since this PR focuses on the FLAX, let's remove them. @wenscarl

Reply:
`variable_with_axes` has been replaced by `self.params`.